### PR TITLE
Bugfix Model Parameter Desync

### PR DESCRIPTION
**Pull request by coreyostrove**
These changes are meant to address the concerns raised in issue #456. To summarize that issue, when manually adding model members of a model's member dictionary it was previously possible to run into scenarios where the values returned for certain public facing attributes, such as the list of parameter labels, had fallen out of sync with the current constituents of the model. In this PR this has been addressed by adding additional checks to various public facing model attributes which check whether we need to rebuild the model, and performs that rebuild if needed before returning the requested value.

This specifically adds checks to the `parameters_labels`, `set_parameter_label`, `parameter_bounds` and `num_modeltest_params` properties/methods. 

Also included is an unrelated bug in the `Circuit` class which got caught by unit testing. 

In the course of testing these changes I also discovered some unexpected and undocumented behavior related to the handling of parameter bounds during the model rebuild stage. When manually setting parameter bounds at the `Model` level and not at the `ModelMember` level these manually specified values get overwritten during the rebuild process by whatever values (if any) are present in the constituent model members. We now have a check in the rebuild process which detects when there is a non-trivial parameter bound vector for a model, and a warning is raised alerting users to this behavior and the possible need to reset any manually specified parameter bounds again at the Model level (if appropriate).

### CODE REVIEW

File: pygsti/models/model.py [hunk 1]
```diff
@@ -610,6 +611,56 @@ def num_params(self):
...
         self._clean_paramvec()
         return len(self._paramvec)
+    
+    @property
+    def parameter_labels(self):
+        """
+        A list of labels, usually of the form `(op_label, string_description)` describing this model's parameters.
+        """
+        self._clean_paramvec()
+        return self._paramlbls
```
Comment (sserita) [inline pygsti/models/model.py:621]: So I thought the merge would catch this but apparently not. If the merge didn't catch it, then I thought the tests would complain about a double property definition. But it looks like neither happened...
But #488 also introduced a parameter_labels, defined on 493 above. Should these be combined like so?

```
@property
    def parameter_labels(self):
        """
        A list of labels, usually of the form `(op_label, string_description)` describing this model's parameters.
        """
        self._clean_paramvec()
        return self._ops_paramlbls_to_model_paramlbls(self._paramlbls)
```

**Review by sserita — COMMENTED** [review]
Potential merge conflict with #488 to resolve (which was not caught by git or the tests, apparently).

File: pygsti/models/model.py (hunk 1 above)
Comment (coreyostrove) [reply pygsti/models/model.py:621]: That is strange... but they should indeed be combined as you did above.

Comment (sserita) [reply pygsti/models/model.py:621]: Mkay, I will do so and merge after that unless you had other blocking issues?

**Review by sserita — APPROVED** [review]

**Pull request merged**